prairielib/lib/migrations.js

```diff
@@ -34,136 +35,272 @@ module.exports.init = function (dir, proj, callback) {
+  for (const migration of migrations) {
+    const { filename, timestamp, index } = migration;
+
+    if (timestamp !== null) {
+      if (seenTimestamps.has(timestamp)) {
+        throw new Error(`Duplicate migration timestamp: ${timestamp} (${filename})`);
+      }
+      seenTimestamps.add(timestamp);
+      hasSeenTimestamp = true;
+    } else {
+      allMigrationsHaveTimestamp = false;
+    }
+
+    if (index !== null) {
+      if (seenIndexes.has(index)) {
+        throw new Error(`Duplicate migration index: ${index} (${filename})`);
+      }
+      seenIndexes.add(index);
+      hasSeenIndex = true;
+    } else {
+      allMigrationsHaveIndex = false;
+    }
+  }
+
+  // Validation: if one migration has a timestamp, *all* migrations must have a timestamp.
+  if (hasSeenTimestamp && !allMigrationsHaveTimestamp) {
+    const filesMissingTimestamp = migrations
+      .filter((m) => m.timestamp === null)
+      .map((m) => m.filename)
+      .join(', ');
+    throw new Error(
+      `The following migration files are missing timestamps: ${filesMissingTimestamp}`
+    );
+  }
+
+  // Validation: if one migration has an index, *all* migrations must have an index.
+  if (hasSeenIndex && !allMigrationsHaveIndex) {
+    const filesMissingIndex = migrations
+      .filter((m) => m.index === null)
+      .map((m) => m.filename)
+      .join(', ');
+    throw new Error(`The following migration files are missing indexes: ${filesMissingIndex}`);
+  }
+
+  return migrations;
+}
+
+/**
+ * Determine the ordering of migrations. We will have already validates that either all migrations
+ * have a timestamp or none of them do; ditto for indexes. So we can safely order on either
+ * one or the other. Default to the timestamp if they're available, otherwise order on the index.
+
+ * @param {MigrationFile[]} migrationFiles
+ * @return {MigrationFile[]}
+ */
+function sortMigrationFiles(migrationFiles) {
+  let allMigrationsHaveTimestamp = migrationFiles.every((m) => m.timestamp);
+  return migrationFiles.sort((a, b) => {
+    if (allMigrationsHaveTimestamp) {
+      return a.timestamp.localeCompare(b.timestamp);
+    } else {
+      return a.index - b.index;
+    }
+  });
+}
+
+/**
+ *
+ * @param {MigrationFile[]} migrationFiles
+ * @param {{ index: number | null, timestamp: string | null }[]} executedMigrations
+ * @return {MigrationFile[]}
+ */
+function getMigrationsToExecute(migrationFiles, executedMigrations) {
+  // If no migrations have ever been run, run them all.
+  if (executedMigrations.length === 0) {
+    return migrationFiles;
+  }
+
+  // If our migration files have timestamps, use them to determine which
+  // have or have not executed. Otherwise, use the index.
+  if (migrationFiles.every((m) => m.timestamp)) {
+    return migrationFiles.filter(
+      (m) => !executedMigrations.some((e) => e.timestamp === m.timestamp)
+    );
+  } else {
+    return migrationFiles.filter((m) => !executedMigrations.some((e) => e.index === m.index));
```

Comment:
Technically this is a change in behavior because we used to filter on `index > maxAppliedIndex`. I don't _think_ that we would ever have ended up in a situation where we skipped some migrations, but I wanted to at least point this out in case I'm missing some other impact of this change.

Reply:
This is actually deliberate - in the case of working with multiple branches, IMO it's a feature not a bug that we'd apply any migrations that got added when we `git merge master` into a working branch.

Reply:
Oh yes, I'm 100% on board for the future behavior working like this. I was only highlighting that it will also apply retroactively (I think?). But I don't think this is a concern in practice.

Reply:
Ah, yeah - if we for some reason have unexecuted migrations lying around, we probably have bigger problems 😂.